### PR TITLE
Feature/predictive slotcar

### PR DESCRIPTION
**Pull request by arjo129**
This PR provides a way to chain colinear waypoints if the next waypoint is collinear and the departure time will already pass by the time it arrives.

Some scripts to test within office world.xml (starting from station1 going to cubicle3): https://gist.github.com/arjo129/1c8bf22a7f802eb081e370273755f79a

### CODE REVIEW

**Review by mxgrey — CHANGES_REQUESTED** [review]
This looks like a great start, but I think there are some logic issues that need to be addressed. Most of them are described in the inline comments below, but there's one that doesn't fit as an inline comment:

The criteria for popping a target off of the queue hasn't changed, when I think it does need to change to be truly robust. Right now a target gets popped off based on the robot's proximity to the target. Since we might not come to a complete stop (or even slow down) at each target, there's no guarantee that any of the simulation time steps will have the robot within that proximity before the robot passes over the target. If the robot is moving too fast and/or the simulation timestep is too large, the robot could feasibly move past the target before popping it from the queue.

I would recommend adjusting the `close_enough` criteria to include a condition where if the effective target is not the current target, then we compute whether the robot will "pass over" the current target within the upcoming timestep. If so, we pop the current target from the queue. You can parameterize the line from the current pose to the effective target in terms of time and then orthographically project the current target onto that line and see if the robot can be expected to reach or pass over that projection within the upcoming time step.

File: building_sim_plugins/building_plugins_common/src/slotcar_common.cpp [hunk 1]
```diff
@@ -313,6 +313,38 @@ std::string to_str(uint32_t type)
   return "UNKNOWN: " + std::to_string(type) + "??";
 }
 
+size_t SlotcarCommon::get_next_noncolinear_target(
```
[the remark this answers is not in the document]
Comment (arjo129) [reply building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:316]: Yes. This is bad nmaming on my part it actually returns the last contiguously colinear target

[the remark this answers is not in the document]
Comment (mxgrey) [reply building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:316]: I'm not sure that's true. It looks to me like the index keeps getting incremented until its value reaches the index of a target that is no longer colinear. So in most cases it seems that it *will* be the next non-colinear target.

[the remark this answers is not in the document]
Comment (mxgrey) [reply building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:316]: I tweaked the `pathological_test.py` script to add even more pathology to it which demonstrates the indexing issue here:

https://gist.github.com/mxgrey/f622bf1d1d51d993359943f45f28788e

The issues you'll see are:
1. The robot comes to an unintended stop in the middle of the lane
2. The robot significantly overshoots one of its waypoints, in some cases crashing into the wall

The intended behavior is to move horizontally from `station_1` to the corner (just before the wall), then backtrack halfway to `station_1`, then return to the corner and finish the original pathological case as usual.

[the remark this answers is not in the document]
Comment (arjo129) [reply building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:316]: Thanks! Will fix the issues.

**Pull request closed without merging**